### ISSUE TITLE
Extremely slow project opening

### ISSUE THREAD

**Issue by nlassaux** [usebruno/bruno#2416]
### I have checked the following:

- [X] I use the newest version of bruno.
- [X] I've searched existing issues and found nothing related to my issue.

### Describe the bug

When loading the collection, Bruno stuck for about 30 to 40sec, stops responding, consumes a ton of resources (120% CPU), and then becomes reactive again. This adds a ton of friction when running some quick debugging.

Intuitively, if bruno loads all 31 files, I can imagine that loading time being close to 5s maybe (not accounting for the UI itself, but from my tests the UI seems -visually- to be loading pretty quickly).

### .bru file to reproduce the bug

Unfortunately I have some API keys in there, is there any way I can help without sharing the whole file? I have 31 requests, with about 5 of them that either use or return a 1MB payload. Maybe 2MB at most for one of them.

### Screenshots/Live demo link

<img width="957" alt="Screenshot 2024-06-05 at 11 30 20" src="https://github.com/usebruno/bruno/assets/1917120/27b6dd80-3435-42cf-92bb-8b4208004aec">
<img width="960" alt="Screenshot 2024-06-05 at 11 30 09" src="https://github.com/usebruno/bruno/assets/1917120/f5bd6879-f587-4a3c-ad1f-32c0d1e7303a">

**Comment by gemiusz** [comment]
I make little tests and have same issue :)

I have prepared Collection for this ticket.

Collection:
[GeMiusz_Bruno_BUGS.zip](https://github.com/user-attachments/files/15596734/GeMiusz_Bruno_BUGS.zip)

On my machine (Windows 11,  12CPU, 16GB RAM, SSD) it take 1min 15s to open this collection.

Below stats from TaskManager:

![image](https://github.com/usebruno/bruno/assets/30750006/03bfd534-d32a-4f00-bcbf-22ccd2dcd3fc)

It take 33s of CPU time, average CPU use 5% and maximum utilisation of RAM was 2 673 812 KB.

I hope that data from my tests and prepared collection will be useful for testing and optimization.

**Comment by Its-treason** [comment]
The culprit here is definitely the .bru file parser:

![Bildschirmfoto vom 2024-06-06 08-26-11](https://github.com/usebruno/bruno/assets/39559178/7d1c69cb-3701-40f0-ad05-7b04deb1ea7c)

But not really sure, if that can be improved, because parsing is handled by a third-party library `ohm-js`, see: https://github.com/usebruno/bruno/blob/main/packages/bruno-lang/v2/src/bruToJson.js

**Comment by gemiusz** [comment]
Maybe it's possible to do "parsing" in threads or in background without hanging application, WDYT?

**Comment by nlassaux** [comment]
I don't know if it's relevant, but it looks like ohm-js is all about usability and not efficiency from what I read on their repo. Here is a benchmark I found to compare with alternatives: https://chevrotain.io/performance/

**Comment by nlassaux** [comment]
@Its-treason If you go down further in that call stack, do you see any of the functions that are defined in that file? https://github.com/usebruno/bruno/blob/main/packages/bruno-lang/v2/src/bruToJson.js

I wonder if list operations, merges, etc. can be optimized if they are being called a gazillion times recursively

**Comment by Its-treason** [comment]
@nlassaux I redid the Perfomance recording. It seems to be mostly internal ohm-js functions being called. I attached the CPU-Profile you can open in Chrome devtools (`Perfomance` -> `Load profile..`) and this is the bruno-lang JS-Bundle i used: https://pastecode.io/s/xyjc3qaw

[bruno_gemiusz_performance.cpuprofile](https://github.com/user-attachments/files/15747723/bruno_gemiusz_performance.cpuprofile)

**Comment by nlassaux** [comment]
Thank you! It does seem to be library-bottlenecked indeed. Testing a thing or two to see if I can improve that

**Comment by nlassaux** [comment]
Trying parsimmon as a parser replacement here: https://github.com/nlassaux/bruno/blob/feature/new-parsimmon-parsing/packages/bruno-lang/v2/src/bruToJsonNew.js

Opening your collection is almost instant @gemiusz 

<img width="701" alt="Screenshot 2024-06-09 at 18 20 53" src="https://github.com/usebruno/bruno/assets/1917120/b104d045-6b18-4591-81fa-46a225c53445">
gemiusz:

https://github.com/usebruno/bruno/assets/1917120/c27c54c0-aa15-406f-8ad0-842a9b031b70


Same for my collection. No more expensive loads anymore.

It doesn't pass all tests yet, but it opens our collections, enough for this test :)

**Comment by gemiusz** [comment]
@nlassaux I wondering if this will resolve issue [[BUG] Poor responsiveness of application when using big collection e.g. 3000 .bru files](https://github.com/usebruno/bruno/issues/2422) too.
I have prepared there collection that can be used for testing.

**Comment by nlassaux** [comment]
Good question! Just checked @gemiusz:
- Loading files is faster (2s vs 4s), but interactions are still pretty slow in the UI
- This change seems to provide crazy improvements for large payloads and small improvements for smaller payloads

**Comment by pottsga** [comment]
Any update on this? I'm experiencing this as well.

**Comment by SPIRIT-00** [comment]
Yep, any news? I have large collection 2796 bru files at now and it grows. A relevant topic for me.

**Comment by nlassaux** [comment]
I don't think the main branch received any update regarding that problem. The parser is the issue, so porting the grammar to a new parser should do the trick. But it's also a critical area, we'll need a ton of tests and all to make sure nothing breaks in the process.

**Comment by Headbucket** [comment]
I have a similar problems with the current version (1.26.0).
I have two requests in my collection with a super large json body (1,5 MB and 9 MB).
Bruno takes about 30-45 seconds to start, when this collection is open.

**Is there a possibility to implement something like "lazy loading"?**

I have the same requests in the VS Code Extension "ThunderClient". ThunderClients loads the requests only when I select them and not before, so VS Code and the Extensions starts quick.

**Comment by lambda-science** [comment]
I have very small collection and bruno takes 2-3 min hanging before working. Is there is a hotfix somehow ? 
I was thinking about getting bruno gold because I like it a lot, but probably not before this is fixed...

**Comment by sreelakshmi-bruno** [comment]
Hi, we have a [beta build](https://github.com/usebruno/bruno-nightly-builds/tree/main?tab=readme-ov-file) for this issue. Please check it out and let me know how it goes. Thanks!

**Comment by lambda-science** [comment]
> Hi, we have a [beta build](https://github.com/usebruno/bruno-nightly-builds/tree/main?tab=readme-ov-file) for this issue. Please check it out and let me know how it goes. Thanks!

In my case going from like 1.34/35 to 1.38.1 solved the issue

**Comment by Headbucket** [comment]
> Hi, we have a [beta build](https://github.com/usebruno/bruno-nightly-builds/tree/main?tab=readme-ov-file) for this issue. Please check it out and let me know how it goes. Thanks!

With version v1.39.0-2025.2.6 I noticed a slight improvement. Thank you for that!
Version 1.38.1 loads projects very slowly for me!

**Test scenario**: I have two collections. One is super small and should load quickly. A second one contains very large JSON files in the body.

With Version 1.38.1 bruno freezes for like two minutes.
With Version 1.39.0-2025.2.6 there seems to be some kind of lazy loading on collection level. That is good.
But the total loading time for the two collections seems to be the same.

I would love some kind of lazy loading on request level! So I could see all the requests right after the start. The request is only loaded when selected (like in ThunderClient).

**Comment by jgiganti** [comment]
I am also having the same issue. I was using 1.35 ish and everything loaded quickly.  I just upgraded and I just get a spinning icon for 30-40 seconds on load and then I can run the tests. I also get errors in the console about missing files

**Comment by ClementValot** [comment]
I upgraded from 1.38.1 to 1.39.0 and, even though the electron app is smooth due to lazy-loading, the time before all the requests can be executed is similar or longer. 

It's a good first step, but I feel like the bottleneck remains the parsing library. 

Is switching the library on someone's plate by any chance ?

**Comment by altaiezior** [comment]
@ClementValot I have implemented the peggy-js just waiting for review from the bruno community, it seems promising for our setup

**Comment by Rohan-Bruno** [comment]
Hi @nlassaux, wanted to follow-up on this to check if you are still facing this issue with the latest version?

**Comment by altaiezior** [comment]
@Rohan-Bruno I did a rough check, seems far better than before. But I will do a re-analysis as well before I could say if everything is working fine. What actually changed though?

**Comment by Rohan-Bruno** [comment]
Hi @altaiezior, we've optimised Bruno's overall performance. Apart from this we've also added a beta feature to Cache collections to reduce app load time (Preferences > Cache).

**Comment by altaiezior** [comment]
Yes we checked that, the major bottleneck was the bru parser and now it has been converted to yaml, that part should have been resolved, so that is good.

If we are not planning to continue with bru anymore we can close https://github.com/usebruno/bruno/pull/5143

**Comment by Rohan-Bruno** [comment]
Hi @altaiezior, since Bruno v3.1, any newly created or imported collections are being stored in .yml format. Bruno will continue to fully support .bru files. However, [OpenCollection YAML](https://docs.usebruno.com/opencollection-yaml/overview) is now the recommended format for new collections.

You can refer to our [blog](https://blog.usebruno.com/making-yaml-the-default-in-bruno-v3.1) for more details.
